### ISSUE TITLE
Need long-vector support

### ISSUE THREAD

**Issue by renkun-ken** [Rdatatable/data.table#3957]
Currently, `data.table` does not support long vectors as columns. Long vector support was not requested much (related: https://github.com/Rdatatable/data.table/issues/2751, https://github.com/Rdatatable/data.table/issues/2340). However, I do need handle data with long vectors more frequently as I step more into the area of high frequency data.

I'm not sure how difficult it is to support long vectors but it would be very nice and much appreciated to put this on the roadmap.

**Comment by 2005m** [comment]
Would anyone know why `data.frame` has a limit in rows in the first place? I couldn't find an answer on R-help.

**Comment by renkun-ken** [comment]
Please see https://stat.ethz.ch/R-manual/R-devel/library/base/html/LongVectors.html.

**Comment by MichaelChirico** [comment]
@renkun-ken I know we have built long vector support in a few places, it might be helpful to make a prioritized list of where it's still not implemented?

**Comment by renkun-ken** [comment]
@MichaelChirico Do we have a list of places where long vectors are supported or could be supported?

**Comment by jangorecki** [comment]
I don't think it make sense to start with anything else than `forder`. AFAIK @mattdowle has that on his radar. Once we will have `forder` working for long vectors, then the rest can follow. We may ended up making an abstraction layer to handle int32/int64 length, because supporting long vectors always may impose performance penalty even for int32 length vectors. It all needs benchmarking to make proper decisions. My suggestion is to hold on with this issue for @mattdowle.

**Comment by aitap** [comment]
@robbyjo, I saw [your patch](https://github.com/Rdatatable/data.table/issues/570#issuecomment-5672748690). I appreciate your need for long data.tables. You're not alone; I would like to see them happen. Maybe Dan Luu is right and [testing, not code review is the future](https://danluu.com/ai-coding/), but for now we the data.table developers still have a duty to our users to actually fit into our heads and understand the code that they run every day. This complicates acceptance of external large patches (a commit history would help!), especially those generated by a tool that doesn't need human understanding to work. Use of zero-length `rownames` attribute for long tables is a bit suspect; proper tests for long vector support will require materialising the ALTREP and working with the large allocations. `int64_t` for in-memory indices is an unfortunate historical choice caused by compiler warnings; `R_xlen_t` expresses the meaning better. `rowIndex`-related helpers that dispatch on type inside loops aren't necessarily great for performance. Partial ALTREP support is risky due to how we use `TRUELENGTH`.

@Roleren, I haven't forgotten about [your comment](https://github.com/Rdatatable/data.table/issues/7438#issuecomment-3785142368). The largest computers I currently have almost unlimited access to in the university have 32G of RAM (and also 32G of relatively fast swap on an NVMe SSD, but that's it). It would be really helpful if you could make a script that would take a toy dataset with $2^{31}-1$ rows and perform all the merges, rolling by-group operations, mutations, et cetera, that you need to work, while fitting in the smallest amount of RAM possible. We could then extend the toy dataset to $2^{31}$ rows (and as much as feasible above that) and try to implement long vector support for all of those, until the script works.

Meanwhile, R-core is [working on native 64-bit integers](https://stat.ethz.ch/pipermail/r-devel/2026-July/084631.html) in base R, so ideally there should be a forward-compatible way to implement that.

**Comment by zachmayer** [comment]
Oooh native 64 bit ints will be nice!

**Comment by Roleren** [comment]
Just wanted to mention I have a working 64 bit data.table fork here:
[data.table link](https://github.com/Roleren/data.table/tree/feature/fst-long-output)
I have an idea that those of us that need 64 bit badly, work together to produce a good product now (outside of the main data.table repo), with extensive testing that must be manually checked by people (no blind ai build).


I started this earlier doing the build manually and now I am using Astra xhigh currently, supported features are:
- Implemented class data.table_long (64bit support for main functions)
- fread, rbindlist both support automatic conversion from 32 to 64 bit row index size
- DT[, sum(score), by = .(col1, col2, col3)] etc supports fast aggregation of tables up to 2^64 rows
- Added 350 new tests for better code coverage and existing tests all pass. 

We could then compare our forks and use the best from all of them. When we believe it is ready (maybe in 2027 ish, we could start discussing a possible merge, lets see what the R guys does with native 64bit support etc). I am actively developing on this patch whenever I have some spare time. 
Possible link topic would be: https://github.com/Rdatatable/data.table/issues/570

Also a note, I have a also made a related fork of fst / fstcore:
[ fst link](https://github.com/Roleren/fst/tree/feature/append-columns)

Which now supports:
- Supports direct conversion from fst to data.table_long (name of the 64bit data.table class), instead of the default base R list output.
- direct appending of columns to disc file! 
- I have started implementing appending of rows to disc file (a bit more tricky, since the last 4k chunk must be zstd recompressed per column)

Also if anyone has new formats with random access for very large tabular data, that is even close to the data.table + fst combo, let me know. I have played a bit with qs2, but it is not even close, and the multi-threading there is much less reliable compared to fst.
